drafts/str4d-overwinter-activation/draft1.rst

```diff
@@ -0,0 +1,196 @@
+
+  It MUST be greater than the value of ``DEPRECATION_HEIGHT`` in the last software version that will not
+  contain support for the network upgrade. It SHOULD be chosen to be reached approximately three months after
+  the first software version containing support for the network upgrade is released.
+
+Activation mechanism
+--------------------
+
+There MUST only be one network upgrade at a time in progress. Concretely, this means that the Zcash blockchain
+is broken into "epochs" of block height intervals ``[ACTIVATION_HEIGHT_{N-1}, ACTIVATION_HEIGHT_N)`` (ie.
+including ``ACTIVATION_HEIGHT_{N-1}`` and exluding ``ACTIVATION_HEIGHT_N``), on which consensus rule sets are
+defined.
+
+Consensus rules themselves (and any network behavior or surrounding code that depends on them) MUST be gated
+by block height checks. For example:
+
+.. code:: cpp
+
+  if (chainActive.Tip()->nHeight >= OVERWINTER.ACTIVATION_HEIGHT) {
+      // Overwinter-specific logic
+  } else {
+      // Pre-Overwinter logic
+  }
+
+  // ...
+
+  if (pindex->nHeight >= OVERWINTER.ACTIVATION_HEIGHT) {
+      // Overwinter consensus rules applied to block
+  } else {
+      // Pre-Overwinter consensus rules applied to block
+  }
+
+
+Block parsing
+`````````````
+Incoming blocks known to have a particular height (due to their parent chain being entirely known) MUST be
+parsed under the consensus rules corresponding to their height.
+
+Incoming blocks with unknown heights (because at least one block in their parent chain is unknown) MUST NOT be
+considered valid, but MAY be cached for future consideration after all their parents have been received.
+
+Chain reorganization
+````````````````````
+It is possible for a reorganization to occur that rolls back from after the activation height, to before that
+height. This can handled in the same way as any regular chain orphaning or reorganization, as long as the new
+chain is valid over the same epochs.
+
+Post-activation upgrading
+`````````````````````````
+If a user does not upgrade their node to a compatible software version before ``ACTIVATION_HEIGHT`` is
+reached, their node will follow the pre-upgrade chain and download blocks that are incompatible with the
+post-upgrade branch. If the user subsequently upgrades their node to a compatible software version, the node
+will consider these blocks to be invalid, and MUST take one of the two following actions:
+
+- Discard all blocks of height ``ACTIVATION_HEIGHT`` and above, and then synchronize with the network.
+
+- Shut down and alert the user of the issue. In this case, the node could offer an option to perform the first
+  action.
+
+Memory pool
+-----------
+
+While the current chain tip height is below ``ACTIVATION_HEIGHT``, nodes SHOULD NOT accept transactions that
+will only be valid on the post-upgrade chain.
+
+When the current chain tip height reaches ``ACTIVATION_HEIGHT``, the node's local transaction memory pool
+SHOULD be cleared.
```

Comment:
When we say "cleared" we really mean cleared of transactions which are not valid for the new set of upgraded rules.  Otherwise, depending on client implementation a node might have a race condition where just as the memory pool is about to be cleared, a valid transaction arrives, only to be evicted.